observatory-dags/observatory/dags/telescopes/geonames.py

```diff
@@ -110,266 +59,196 @@ def first_sunday_of_month(datetime: Pendulum) -> Pendulum:
-        """ Download release to file.
-        If develop environment, copy test file from this repository to the right location. Else download from url.
+    def transform(self):
+        """ Transforms release by storing file content in gzipped csv format.
 
-        :param kwargs: the context passed from the PythonOperator. See
-        https://airflow.apache.org/docs/stable/macros-ref.html
-        for a list of the keyword arguments that are passed to this argument.
-        :return: None.
+        :return: None
         """
 
-        # Get variables
-        environment = Variable.get(AirflowVars.ENVIRONMENT)
+        with open(self.extract_path, 'rb') as file_in:
+            with gzip.open(self.transform_path, 'wb') as file_out:
+                shutil.copyfileobj(file_in, file_out)
 
-        # Fetch release date and create GeonamesRelease object
-        release_date = fetch_release_date()
-        release = GeonamesRelease(release_date)
 
-        # Download release
-        if environment == 'test':
-            shutil.copy(test_data_path(), release.filepath_download)
-        else:
-            download_release(release)
+class GeonamesTelescope(SnapshotTelescope):
+    """
+    A Telescope that harvests the GeoNames geographical database: https://www.geonames.org/
 
-        # Push release date for other tasks
-        ti: TaskInstance = kwargs['ti']
-        ti.xcom_push(GeonamesTelescope.RELEASES_TOPIC_NAME, release, kwargs['execution_date'])
+    Saved to the BigQuery table: <project_id>.geonames.geonamesYYYYMMDD
+    """
 
-    @staticmethod
-    def upload_downloaded(**kwargs):
-        """ Task to upload the downloaded Geonames release.
+    DAG_ID = 'geonames'
+    DATASET_DESCRIPTION = 'The GeoNames geographical database: https://www.geonames.org/'
+
+    def __init__(self, dag_id: str = DAG_ID, start_date: datetime = datetime(2020, 9, 1),
+                 schedule_interval: str = '@weekly', dataset_id: str = 'geonames',
+                 catchup: bool = False):
+
+        airflow_vars = [AirflowVars.DATA_PATH, AirflowVars.PROJECT_ID, AirflowVars.DATA_LOCATION,
+                        AirflowVars.DOWNLOAD_BUCKET, AirflowVars.TRANSFORM_BUCKET]
+        super().__init__(dag_id, start_date, schedule_interval, dataset_id,
+                         source_format=SourceFormat.CSV,
+                         dataset_description=self.DATASET_DESCRIPTION,
+                         load_bigquery_table_kwargs={'csv_field_delimiter': '\t', 'csv_quote_character': ''},
+                         catchup=catchup, airflow_vars=airflow_vars)
+        self.add_setup_task(self.check_dependencies)
+        self.add_setup_task(self.list_releases)
+        self.add_task(self.download)
+        self.add_task(self.upload_downloaded)
+        self.add_task(self.extract)
+        self.add_task(self.transform)
+        self.add_task(self.upload_transformed)
+        self.add_task(self.bq_load)
+        self.add_task(self.cleanup)
```

Comment:
> I noticed the construction of Telescopes is slightly different for the telescopes and even though it's a small detail I think it would be nice to be consistent.
> 
> * In this case there is a DAG_ID class variable which is passed on to the dag_id parameter.
> * The DATASET_DESCRIPTION is a class variable, but passed on to the dataset_description parameter in the super call .
> * airflow_vars is a variable in the init method (rather than being a parameter)
> * Some values are directly defined in the super() init call, like 'source_format' and 'load_bigquery_table_kwargs'. Rather than being parameters in the child init or (class) variables.
> 
> When constructing a telescope for the dag (see below), which properties of the telescope do we want to be able to configure during this process?
> 
> ```
> from observatory.dags.telescopes.grid import GridTelescope
> 
> telescope = GridTelescope()
> globals()[telescope.dag_id] = telescope.make_dag()
> ```
> 
> In this example you could create the telescope and fill in the start_date, schedule_interval, dataset_id and catchup, but not the dag_id, dataset_description, airflow_vars or bigquery_table_kwargs. It seems that in practice (so far) we don't configure any of these, but leave them to the default values.

Yeah it would be good to make it consistent.

We will want to be able to customise the dag_id parameter, because we will create multiple DAG instances (with different ids) with the new Mellon telescopes, based on what the REST API returns.

The class variable DAG_ID (as opposed to self.dag_id) is still required for all telescopes as they are used by the [doi.py DAG](https://github.com/The-Academic-Observatory/observatory-platform/blob/develop/observatory-dags/observatory/dags/dags/doi.py#L73-L108).

> 
> If we will always leave them to the default values I would prefer to list the parameters as variables in the init method (like airflow_vars), just because it looks a bit cleaner to me having them below each other compared to on 1 line as parameters.
> 
> I'm just wondering what the reasoning is behind which properties are parameters/which are not and if we should make this consistent across telescopes.

My thinking was that the users wouldn't need to change certain parameters in the final telescopes (e.g. GeonamesTelescope), e.g. source_format, load_bigquery_table_kwargs, airflow_vars etc. I'm happy to make them all keyword arguments in the constructor if that makes them all consistent though.

Reply:
> The class variable DAG_ID (as opposed to self.dag_id) is still required for all telescopes as they are used by the doi.py DAG.

Thanks, I did not realise this, but good to know.

> My thinking was that the users wouldn't need to change certain parameters in the final telescopes (e.g. GeonamesTelescope), e.g. source_format, load_bigquery_table_kwargs, airflow_vars etc. I'm happy to make them all keyword arguments in the constructor if that makes them all consistent though.

Sounds good :-)